### ISSUE TITLE
query_duration_ms unstable 

### ISSUE THREAD

**Issue by chenglinli0** [ClickHouse/ClickHouse#60345]
(you don't have to strictly follow this form)

**Describe the situation**
What exactly works slower than expected?

In production env, we have a schedule task that running clickhosue sql 1 time / minute. 

the task has 3 replicas , so the clickhouse server will get 3 sql in every minute.

but we found the sql cost time is unstable , most of them are between 6s -10s.  and there is some time that the time become very high.

here is the detail distribution , every point means a sql runing .

![image](https://github.com/ClickHouse/ClickHouse/assets/113302886/d7207c07-073b-42fa-8741-808a24d719c5)

is there anyway that find the root cause?

**How to reproduce**
* Which ClickHouse server version to use
23.8.9
* Which interface to use, if matters
official jdbc-driver with http 
* Non-default settings, if any
4 nodes including 2 share and 2 replicas 

* `CREATE TABLE` statements for all tables involved
* 
* Sample data for all these tables, use [clickhouse-obfuscator](https://github.com/ClickHouse/ClickHouse/blob/master/programs/obfuscator/Obfuscator.cpp#L42-L80) if necessary
* Queries to run that lead to slow performance

**Expected performance**
the query_duration_ms  should be more stable

**Additional context**

**Comment by qoega** [comment]
You can check ProfileEvents in query_log to see which part of execution took more time. It can be anything.

**Comment by chenglinli0** [comment]
> You can check ProfileEvents in query_log to see which part of execution took more time. It can be anything.

here are two querys's profile_events and their query_duration_ms.

the first one has more events('FileOpen=542, OpenedFileCacheMisses=542') than the second one. 

query_duration_ms:6601

[{Query=1, SelectQuery=1, QueriesWithSubqueries=23, SelectQueriesWithSubqueries=23, FileOpen=542, ReadBufferFromFileDescriptorReadBytes=621359635, ReadCompressedBytes=540513198, CompressedReadBufferBlocks=82365, CompressedReadBufferBytes=5693197195, OpenedFileCacheHits=51867, OpenedFileCacheMisses=542, OpenedFileCacheMicroseconds=127214, IOBufferAllocs=104829, IOBufferAllocBytes=4043480968, ArenaAllocChunks=2227, ArenaAllocBytes=54321152, FunctionExecute=46276, MarkCacheHits=52409, CreatedReadBufferOrdinary=52409, DiskReadElapsedMicroseconds=784648, NetworkReceiveElapsedMicroseconds=18449, NetworkSendElapsedMicroseconds=141, NetworkReceiveBytes=2059238, NetworkSendBytes=11033, DistributedConnectionTries=4, DistributedConnectionUsable=4, SuspendSendingQueryToShard=4, SelectedParts=51, SelectedRanges=1211, SelectedMarks=1485, SelectedRows=15197539, SelectedBytes=2951718814, WaitMarksLoadMicroseconds=550455, ContextLock=55282, ContextLockWaitMicroseconds=901172, RWLockAcquiredReadLocks=39, RWLockReadersWaitMilliseconds=4, PartsLockHoldMicroseconds=154, RealTimeMicroseconds=3243183336, UserTimeMicroseconds=13667884, SystemTimeMicroseconds=5679689, SoftPageFaults=996909, OSCPUWaitMicroseconds=322396, OSCPUVirtualTimeMicroseconds=19305813, OSReadChars=621353984, OSWriteChars=1321984, QueryProfilerRuns=3260, ThreadPoolReaderPageCacheMiss=79608, ThreadPoolReaderPageCacheMissBytes=621359635, ThreadPoolReaderPageCacheMissElapsedMicroseconds=784648, AsynchronousReadWaitMicroseconds=367537672}]

query_duration_ms:18269
[{Query=1, SelectQuery=1, QueriesWithSubqueries=23, SelectQueriesWithSubqueries=23,               ReadBufferFromFileDescriptorReadBytes=614248498, ReadCompressedBytes=527945536, CompressedReadBufferBlocks=80703, CompressedReadBufferBytes=5502248730, OpenedFileCacheHits=51694,                            OpenedFileCacheMicroseconds=113309, IOBufferAllocs=103398, IOBufferAllocBytes=4034146013, ArenaAllocChunks=2247, ArenaAllocBytes=55873536, FunctionExecute=45979, MarkCacheHits=51694, CreatedReadBufferOrdinary=51694, DiskReadElapsedMicroseconds=727208, NetworkReceiveElapsedMicroseconds=42148, NetworkSendElapsedMicroseconds=182, NetworkReceiveBytes=2223016, NetworkSendBytes=11033, DistributedConnectionTries=4, DistributedConnectionUsable=4, SuspendSendingQueryToShard=4, SelectedParts=51, SelectedRanges=1207, SelectedMarks=1487, SelectedRows=15189499, SelectedBytes=2958885430, WaitMarksLoadMicroseconds=483041, ContextLock=54595, ContextLockWaitMicroseconds=399344, RWLockAcquiredReadLocks=39,                                  PartsLockHoldMicroseconds=136, RealTimeMicroseconds=9776258067, UserTimeMicroseconds=14254719, SystemTimeMicroseconds=5057595, SoftPageFaults=778288, OSCPUWaitMicroseconds=283164, OSCPUVirtualTimeMicroseconds=19270861, OSReadChars=614250496, OSWriteChars=2253824, QueryProfilerRuns=9786, ThreadPoolReaderPageCacheMiss=77731, ThreadPoolReaderPageCacheMissBytes=614248498, ThreadPoolReaderPageCacheMissElapsedMicroseconds=727208, AsynchronousReadWaitMicroseconds=340063572}]

**Comment by jetgm** [comment]
I hit same problem. One sql cost time different on different node,  and different time.
It seems that  the ProfileEvents value of query_log  is not correctly, if the sql duration is long, I cannot use ProfileEvents to judge which is the root reason from CPU,MEM,DISK,NETWORK

**Comment by chenglinli0** [comment]
For this case, I have solved the problem.

The most influential factor is QPS.

Clickhouse seems not good at  high concurrent query. 

In my case, when QPS become more than 100 reqs/s, the query_duration_ms  become high.

1. give clickhosue more cpus.
2. reduce QPS
3. limit user settings   <max_threads>4</max_threads>

It can help.

and another question.

How  clickhosue allocate cpu core to handle querys? what if a sql is quite simple like 'select 1'. 

we found a lot of query like this.It's mostly generated by 'jdbc driver' ping request.
